src/graphics.rs

```diff
@@ -1628,6 +1634,263 @@ impl Drawable for Sprite {
+                let bb = glyph.pixel_bounding_box().unwrap_or(inval_bb);
+
+                char_map.insert(
+                    c,
+                    CharInfo {
+                        vertical_offset: bb.min.y,
+                        uvs: Box2::new(
+                            texture_cursor.0 as f32 / texture_width as f32,
+                            texture_cursor.1 as f32 / texture_height as f32,
+                            bb.width() as f32 / texture_width as f32,
+                            bb.height() as f32 / texture_height as f32,
+                        ),
+                        width: glyph.unpositioned().h_metrics().advance_width,
+                    },
+                );
+
+                glyph.draw(|x, y, v| {
+                    let x: u32 = texture_cursor.0 as u32 + x;
+                    let y: u32 = texture_cursor.1 as u32 + y;
+                    let c = (v * 255.0) as u8;
+                    let color = Rgba([255, 255, 255, c]);
+                    texture.put_pixel(x, y, color);
+                });
+
+                texture_cursor.0 += bb.width() as usize + MARGIN;
+            }
+            texture_cursor.1 += height + MARGIN;
+            texture_cursor.0 = 0;
+        }
+
+        texture.save("font_atlas.png")?;
+
+        Ok(FontAtlas {
+            font_map: char_map,
+            font_texture: Cached::new(Texture::from_rgba8(
+                ctx,
+                texture_width as u16,
+                texture_height as u16,
+                &texture,
+            )),
+        })
+    }
+
+    fn get_char_list(char_list_type: CharacterListType) -> Result<Vec<char>> {
+        let char_list = match char_list_type {
+            CharacterListType::AsciiSubset => [0x20..0x7F].iter(),
+            CharacterListType::Ascii => [0x00..0xFF].iter(),
+            CharacterListType::Cyrillic => [
+                0x0020u32..0x00FF, // Basic Latin + Latin Supplement
+                0x0400u32..0x052F, // Cyrillic + Cyrillic Supplement
+                0x2DE0u32..0x2DFF, // Cyrillic Extended-A
+                0xA640u32..0xA69F, // Cyrillic Extended-B
+            ]
+            .iter(),
+            CharacterListType::Thai => [
+                0x0020u32..0x00FF, // Basic Latin
+                0x2010u32..0x205E, // Punctuations
+                0x0E00u32..0x0E7F, // Thai
+            ]
+            .iter(),
+
+            CharacterListType::Vietnamese => [
+                0x0020u32..0x00FF, // Basic Latin
+                0x0102u32..0x0103,
+                0x0110u32..0x0111,
+                0x0128u32..0x0129,
+                0x0168u32..0x0169,
+                0x01A0u32..0x01A1,
+                0x01AFu32..0x01B0,
+                0x1EA0u32..0x1EF9,
+            ]
+            .iter(),
+            CharacterListType::Chinese => return Err(anyhow!("Chinese fonts not yet supported")),
+            CharacterListType::Japanese => return Err(anyhow!("Japanese fonts not yet supported")),
```

Comment:
prefer `bail!(...)` to `return Err(anyhow!(...))` they do the same thing but one is more concise and doesn't explicitly construct any error type.